automation/run/run.go

```diff
@@ -54,23 +75,42 @@ const (
 // a transform. State structs can act as a sink of transform events, collapsing
 // the state transition of multiple transform events into a single structure
 type State struct {
-	ID        string       `json:"id"`
-	Number    int          `json:"number"`
-	Status    Status       `json:"status"`
-	Message   string       `json:"message"`
-	StartTime *time.Time   `json:"startTime"`
-	StopTime  *time.Time   `json:"stopTime"`
-	Duration  int          `json:"duration"`
-	Steps     []*StepState `json:"steps"`
-}
-
-// NewState is a simple constructor to remind package consumers that state
-// structs must be initialized with an identifier to act as a sink of transform
-// events
-func NewState(id string) *State {
-	return &State{
-		ID: id,
+	ID         string       `json:"id"`
+	WorkflowID workflow.ID  `json:"workflowID"`
+	Number     int          `json:"number"`
+	Status     Status       `json:"status"`
+	Message    string       `json:"message"`
+	StartTime  *time.Time   `json:"startTime"`
+	StopTime   *time.Time   `json:"stopTime"`
+	Duration   int          `json:"duration"`
+	Steps      []*StepState `json:"steps"`
+}
+
+// Validate errors if the run is not valid
+func (rs *State) Validate() error {
+	if rs.ID == "" {
+		return ErrNoID
+	}
+	if rs.WorkflowID.String() == "" {
+		return ErrNoWorkflowID
+	}
+	return nil
+}
+
+// Copy shallowly copies the contents of run parameter into the receiver
+func (rs *State) Copy(run *State) {
```

Comment:
The function names feels off, usually `Copy` makes a copy of the current thing and returns it, in order to be used like this:

```
a := b.Copy()
```

Could we rename this to `CopyFrom`?

Reply:
I'd rather just change it to be the more correct version of `func (rs *State) Copy() State`!

ty